### ISSUE TITLE
Consider breaking down `AppDeployToolkitMain.ps1` into smaller files.

### ISSUE THREAD

**Issue by mjr4077au** [PSAppDeployToolkit/PSAppDeployToolkit#888]
**Is your feature request related to a problem? Please describe.**
The file `AppDeployToolkitMain.ps1` is nothing short of gargantuan these days, at 16,000+ lines of code. This file consists of functions for internal/private use, as well as public functions for end users to call.

**Describe the solution you'd like**
I'd like to see internal/private functions move to `AppDeployToolkitMainPrivate.ps1`, and public functions move to `AppDeployToolkitMainPublic.ps1`. Each of these files can be dot-sourced into `AppDeployToolkitMain.ps1` as required.

I think this would drastically improve readability of the code, and help end users know what they're supposed to be playing with instead of having private functions intermixed with public ones.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by bulax** [comment]
For private functions I believe you should use "AppDeployToolkitExtensions.ps1" This is place where you should put all your code. If you want to make it more complex, you can create your module and import it inside Deploy-Application.ps1

**Comment by mjr4077au** [comment]
@bulax sorry mate, when I said "private", I meant PSADT functions that explicitly state "This is an internal script function and should typically not be called directly.". I should have used "internal" as it would have possibly been clearer.

**Comment by mjr4077au** [comment]
Well, I did officially become a member of this team the other day, so I could definitely make a start 😊. I'm not touching anything until v4.0 is officially roadmapped and agreed by all parties however.

**Comment by djalexnl** [comment]
Let's keep it as it is; in my opinion, it's not that hard to understand. Moreover, maintaining a fork for my company is easier this way. Splitting this into multiple files would only complicate matters. It doesn't resolve the issue. As an official member, I hope you have the ability to listen to the end users.

**Comment by mjr4077au** [comment]
@djalexnl just curious about what's required you to maintain your own fork of PSADT? Is there anything that could be incorporated to avoid this? I too started off maintaining my own fork because reasons but the team has been very forthcoming with a number of things I've submitted.

**Comment by djalexnl** [comment]
> @djalexnl just curious about what's required you to maintain your own fork of PSADT? Is there anything that could be incorporated to avoid this? I too started off maintaining my own fork because reasons but the team has been very forthcoming with a number of things I've submitted.

I've been doing this work for years and had my own deployment toolkit in VBScript since 2016. At our company, the user is always central, including communication towards the end user. We opt to inform the user if something goes awry with the installation, a choice not made here. For example, if an installation fails with an error 1603, our users receive a clear notification about the issue and instructions on contacting our help desk. I never use Throw and would never want a script to crash, leaving a progress screen spinning at the user's workstation while they wait for the installation to complete. For instance, I've now added a throwexit function. For all actions in the toolkit where Throw is used, I resort to my own throwexit function, subsequently closing the toolkit neatly via the exit-script function. Users are promptly informed that the installation failed, with the exit code provided for reference. I also log extensively around the installations: what happens during, before, after, and the delta, so we can quickly ascertain what went wrong. When there are blocking issues or pending files, users are neatly informed that the workstation requires a restart. If MSI engines are in use and take longer than a specified period, users are notified that the installation cannot proceed yet and should try again later. We seek much more interaction with the user. But let me emphasize that I greatly appreciate the toolkit and the effort invested in it. I strive to keep as much code intact as possible. However, when you are serving a vast number of users, it's essential to involve the users more in the process; otherwise, they continue working on corrupt machines. Additionally, I want to create a sort of installation loop that I've utilized in the past. For example, if you install an application and it fails, that's okay; it happens, especially when updating thousands of clients. Often, when analyzing log files, you identify a trend in why, for instance, 10 percent of installations fail. Generally, the root cause becomes apparent only after executing the implementation. To address this, I've devised this installation loop. It starts the installation, checks the exit code—if it's null, great, the process exits. If not, it moves into a module or ps1 script. If there's nothing further, it exits with an error code, and the user is notified of the failure. However, in the module or ps1, I can specify, for example, to check the logs for a specific line. If present, something is fixed or adjusted on the workstation, and then the loop is re-entered to restart the installation, with a maximum number of attempts and a timeout. The user sees a message that the installation may take longer. This method has allowed many of our applications to have an auto-repair function. I've used this in the past, and it has provided us with significant advantages. When we handle many complex implementations, this is one of the key benefits. We deploy granularly, so the trend of the implementation becomes apparent very early. If Version 1 has a 10 percent dropout rate, we address it with a fix, so Version 2 has only a 4 percent dropout, and Version 3 drops to 1 percent, and so on.

**Comment by mjr4077au** [comment]
With your `throwexit` function, I don't see why something couldn't throw and be caught in Deploy-Application.ps1's main try/catch block. From the outset, it sounds like its doing the same thing - that is, a function throws and it keeps throwing through the callstack until something catch's it, such as that main try/catch block, which does call `Exit-Script` and exits out gracefully. There's no real need to exit out immediately at the exception point, and the main try/catch block will ensure the script is never terminated or crashes without a graceful cleanup.

Below is an example of our company's Deploy-Application.ps1 script. I change a few variables that Main sets up as I don't like how they are, and you can't really see what the `Invoke-*` functions we've got stuffed in `AppDeployToolkitExtensions.ps1` is doing, but ultimately I'm calling `Execute-Process` with `-PassThru`. If the execution bombed out, I add a new PSNoteProperty to the PassThru data and return it to the main callstack. Here it checks if it failed or not and provides a message dialogue to the user. If there's an actual throw, of course its caught in the catch block and a proper cleanup occurs. This is effectively on a vanilla PSADT backend (once my PRs are merged).

```PowerShell
# Commence PSADT operation.
try
{
	# Initialise PSADT.
	. "$PSScriptRoot\AppDeployToolkit\AppDeployToolkitMain.ps1"
	$deployModeSilent = $deployMode -eq 'Silent'
	$installTitle = ($installTitle -replace "(^$appVendor$(if ($AppName.Equals('Google Chrome') -or $AppName.Contains($AppVersion)) {"|$AppVersion$"}))").Trim()
	$mainErrorPreamble = $mainErrorPreamble -replace '(^This deployment)', $installTitle

	# Ensure the deployment app variable is valid before continuing or presenting anything to the user.
	# This could be invalid if a bad filename or script command/function was provided by the caller.
	if (![System.IO.File]::Exists($deployAppVariable.Value) -and ($deployAppVariable.Value -notmatch $uuidRegex) -and !(Microsoft.PowerShell.Core\Get-Command -Name $deployAppVariable.Value -ErrorAction Ignore))
	{
		throw "The supplied $deployAppParameterSetName value is invalid, unable to continue."
	}

	# Present any dialogs if we're non-silent.
	$installPhase = "Pre-$DeploymentType"
	[System.Void](Show-InstallationWelcome @siwParams)
	[System.Void](Show-InstallationProgress -NoRelocation)

	# Commence deployment operation.
	$installPhase = $DeploymentType
	$instData = if ($useDefaultMsi)
	{
		Invoke-MsiOperation
	}
	elseif ($PSCmdlet.ParameterSetName.Equals('Script'))
	{
		Invoke-ScriptOperation
	}
	elseif ($PSCmdlet.ParameterSetName.Equals('Executable'))
	{
		Invoke-ExecutableOperation
	}
	else
	{
		throw "Error occurred while determining how to invoke the requested operation. Please report this issue for further diagnosis."
	}

	# Process deployment results and exit out.
	$installPhase = "Post-$DeploymentType"
	if ($instData.PSObject.Properties.Name.Contains('DialogMessage'))
	{
		Show-DialogBox -Text "$mainErrorPreamble$($instData.DialogMessage)" -Icon 'Stop' | & (Microsoft.PowerShell.Core\Get-Command -Name Out-Null -Module Microsoft.PowerShell.Core -ErrorAction Stop)
	}
	elseif ($DeployMode -eq 'Interactive')
	{
		Show-InstallationPrompt -Message "Successfully $($DeploymentType.ToLower())ed $installTitle." -ButtonRightText OK -Icon Information -NoWait
	}
	Remove-DownloadedContent
	Exit-Script -ExitCode $instData.ExitCode -ValidExitCodes $ValidExitCodes
}
catch
{
	# We should really only get here if PSADT fails to initialise properly or something goes really bad...
	Write-Log -Message ($mainErrorMessage = "$(Resolve-Error)") -Severity 3 -Source $deployAppScriptFriendlyName
	Show-DialogBox -Text "$mainErrorPreamble$mainErrorMessage" -Icon 'Stop' | & (Microsoft.PowerShell.Core\Get-Command -Name Out-Null -Module Microsoft.PowerShell.Core -ErrorAction Stop)
	Exit-Script -ExitCode 60001 -ValidExitCodes $ValidExitCodes
}
```
![image](https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/assets/48643140/d79fbbcf-66e0-44d2-8986-64cd3d3903b6)

The other stuff sounds interesting but I imagine its a little hard to generalise which is why you've opted for a fork. There's probably specific reasons why your apps may fail at times so needless to say you'd have specific code to handle such exit codes and whatnot. In saying that, such a loop probably doesn't necessitate changes in AppDeployToolkitMain.ps1 unless the default setup for things like Execute-Process isn't meeting requirements.

**Comment by djalexnl** [comment]
> With your `throwexit` function, I don't see why something couldn't throw and be caught in Deploy-Application.ps1's main try/catch block. From the outset, it sounds like its doing the same thing - that is, a function throws and it keeps throwing through the callstack until something catch's it, such as that main try/catch block, which does call `Exit-Script` and exits out gracefully. There's no real need to exit out immediately at the exception point, and the main try/catch block will ensure the script is never terminated or crashes without a graceful cleanup.
> 
> Below is an example of our company's Deploy-Application.ps1 script. I change a few variables that Main sets up as I don't like how they are, and you can't really see what the `Invoke-*` functions we've got stuffed in `AppDeployToolkitExtensions.ps1` is doing, but ultimately I'm calling `Execute-Process` with `-PassThru`. If the execution bombed out, I add a new PSNoteProperty to the PassThru data and return it to the main callstack. Here it checks if it failed or not and provides a message dialogue to the user. If there's an actual throw, of course its caught in the catch block and a proper cleanup occurs. This is effectively on a vanilla PSADT backend (once my PRs are merged).
> 
> ```powershell
> # Commence PSADT operation.
> try
> {
> 	# Initialise PSADT.
> 	. "$PSScriptRoot\AppDeployToolkit\AppDeployToolkitMain.ps1"
> 	$deployModeSilent = $deployMode -eq 'Silent'
> 	$installTitle = ($installTitle -replace "(^$appVendor$(if ($AppName.Equals('Google Chrome') -or $AppName.Contains($AppVersion)) {"|$AppVersion$"}))").Trim()
> 	$mainErrorPreamble = $mainErrorPreamble -replace '(^This deployment)', $installTitle
> 
> 	# Ensure the deployment app variable is valid before continuing or presenting anything to the user.
> 	# This could be invalid if a bad filename or script command/function was provided by the caller.
> 	if (![System.IO.File]::Exists($deployAppVariable.Value) -and ($deployAppVariable.Value -notmatch $uuidRegex) -and !(Microsoft.PowerShell.Core\Get-Command -Name $deployAppVariable.Value -ErrorAction Ignore))
> 	{
> 		throw "The supplied $deployAppParameterSetName value is invalid, unable to continue."
> 	}
> 
> 	# Present any dialogs if we're non-silent.
> 	$installPhase = "Pre-$DeploymentType"
> 	[System.Void](Show-InstallationWelcome @siwParams)
> 	[System.Void](Show-InstallationProgress -NoRelocation)
> 
> 	# Commence deployment operation.
> 	$installPhase = $DeploymentType
> 	$instData = if ($useDefaultMsi)
> 	{
> 		Invoke-MsiOperation
> 	}
> 	elseif ($PSCmdlet.ParameterSetName.Equals('Script'))
> 	{
> 		Invoke-ScriptOperation
> 	}
> 	elseif ($PSCmdlet.ParameterSetName.Equals('Executable'))
> 	{
> 		Invoke-ExecutableOperation
> 	}
> 	else
> 	{
> 		throw "Error occurred while determining how to invoke the requested operation. Please report this issue for further diagnosis."
> 	}
> 
> 	# Process deployment results and exit out.
> 	$installPhase = "Post-$DeploymentType"
> 	if ($instData.PSObject.Properties.Name.Contains('DialogMessage'))
> 	{
> 		Show-DialogBox -Text "$mainErrorPreamble$($instData.DialogMessage)" -Icon 'Stop' | & (Microsoft.PowerShell.Core\Get-Command -Name Out-Null -Module Microsoft.PowerShell.Core -ErrorAction Stop)
> 	}
> 	elseif ($DeployMode -eq 'Interactive')
> 	{
> 		Show-InstallationPrompt -Message "Successfully $($DeploymentType.ToLower())ed $installTitle." -ButtonRightText OK -Icon Information -NoWait
> 	}
> 	Remove-DownloadedContent
> 	Exit-Script -ExitCode $instData.ExitCode -ValidExitCodes $ValidExitCodes
> }
> catch
> {
> 	# We should really only get here if PSADT fails to initialise properly or something goes really bad...
> 	Write-Log -Message ($mainErrorMessage = "$(Resolve-Error)") -Severity 3 -Source $deployAppScriptFriendlyName
> 	Show-DialogBox -Text "$mainErrorPreamble$mainErrorMessage" -Icon 'Stop' | & (Microsoft.PowerShell.Core\Get-Command -Name Out-Null -Module Microsoft.PowerShell.Core -ErrorAction Stop)
> 	Exit-Script -ExitCode 60001 -ValidExitCodes $ValidExitCodes
> }
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/48643140/320268886-d79fbbcf-66e0-44d2-8986-64cd3d3903b6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI0ODA0NTcsIm5iZiI6MTcxMjQ4MDE1NywicGF0aCI6Ii80ODY0MzE0MC8zMjAyNjg4ODYtZDc5ZmJiY2YtNjZlMC00NGQyLTg5ODYtNjRjZDNkMzkwM2I2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA3VDA4NTU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNiMjE4ZTU1ZGQwZDNiMzRlODhhZGFhZDNkMWY2YjFjYzI4YTJlNzVlZWMxNDRjNTFmNDU0ZmY4NjUwMWRlY2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cQaINAyo6uJLtJBmIPE2Eh9B7_qx66s-fTEahPX7PxA)
> 
> The other stuff sounds interesting but I imagine its a little hard to generalise which is why you've opted for a fork. There's probably specific reasons why your apps may fail at times so needless to say you'd have specific code to handle such exit codes and whatnot. In saying that, such a loop probably doesn't necessitate changes in AppDeployToolkitMain.ps1 unless the default setup for things like Execute-Process isn't meeting requirements.

I understand what you're saying, and there are certainly different methods to approach things. I try to adhere as much as possible to the original code while also implementing the changes and maintenance that are suggested. I closely follow the development and, let me be very clear, I think it's a fantastic toolkit. It's wonderful to see how we can collectively enhance the implementation for the end-user, and I greatly appreciate everyone's input in this process. With the current toolkit, I can easily keep everything up to date while also maintaining my own functionality. Thank you for your feedback, understanding, and for sharing your thoughts and approaches.

**Comment by mjr4077au** [comment]
No worries at all! Knowledge and ideas are there to be shared and I've appreciated hearing your approach also. If you ever feel you've got something that's worthy of a PR, or would be easier if upstreamed, definitely don't be afraid to send it in. I've found everyone on the PSADT team to be wonderful to work with and everyone's genuinely here to make the best toolkit we possibly can. Hell, a few PRs later and I even got myself a job!
